### PR TITLE
Make DaoManager concurrent

### PR DESCRIPTION
**Pull request by MarcMil**
Removing the bottlenecks introduced by synchronized using concurrent hashmaps.

### CODE REVIEW

File: src/main/java/com/j256/ormlite/dao/BaseDaoImpl.java [hunk 1]
```diff
@@ -57,7 +57,7 @@
  *            needs an ID parameter however so you can use Void or Object to satisfy the compiler.
  * @author graywatson
  */
-public abstract class BaseDaoImpl<T, ID> implements Dao<T, ID> {
+public class BaseDaoImpl<T, ID> implements Dao<T, ID> {
```
Comment (noordawod) [inline src/main/java/com/j256/ormlite/dao/BaseDaoImpl.java:60]: Why was this change introduced? It doesn't appear that you require this change for your pr, so I suggest to undo.

File: src/main/java/com/j256/ormlite/dao/DaoManager.java [hunk 2]
```diff
@@ -96,16 +97,14 @@ public class DaoManager {
 			}
 		}
 
-		registerDao(connectionSource, dao);
-		@SuppressWarnings("unchecked")
```
Comment (ndawod) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:100]: Keep the suppression please, this is meant for IDEs/linters.

File: src/main/java/com/j256/ormlite/dao/DaoManager.java [hunk 3]
```diff
@@ -277,35 +263,21 @@ private static void removeDaosFromConnectionClassMap(ConnectionSource connection
...
-	private static void addDaoToTableMap(TableConfigConnectionSource key, Dao<?, ?> dao) {
-		if (tableConfigMap == null) {
-			tableConfigMap = new HashMap<TableConfigConnectionSource, Dao<?, ?>>();
+	private static Dao<?, ?> addDaoToTableMap(TableConfigConnectionSource key, Dao<?, ?> dao) {
+		Dao<?, ?> old = tableConfigMap.putIfAbsent(key, dao);
+		if (old != null) {
+			return old;
+		} else {
+			return dao;
 		}
```
Comment (ndawod) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:272]:
```suggestion
		if (old != null) {
			return old;
		}

		return dao;
```

File: src/main/java/com/j256/ormlite/dao/DaoManager.java [hunk 4]
```diff
@@ -213,57 +213,43 @@ public static synchronized void registerDaoWithTableConfig(ConnectionSource conn
...
-	private static void addDaoToClassMap(ClassConnectionSource key, Dao<?, ?> dao) {
-		if (classMap == null) {
-			classMap = new HashMap<ClassConnectionSource, Dao<?, ?>>();
+	private static Dao<?, ?> addDaoToClassMap(ClassConnectionSource key, Dao<?, ?> dao) {
+		Dao<?, ?> old = classMap.putIfAbsent(key, dao);
+		if (old != null) {
+			return old;
+		} else {
+			return dao;
 		}
```
Comment (ndawod) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:248]:
```suggestion
		if (old != null) {
			return old;
		}
		
		return dao;
```

Comment (noordawod) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:248]:
```suggestion
		if (old != null) {
			return old;
		}
		
		return dao;
```

File: src/main/java/com/j256/ormlite/dao/DaoManager.java (hunk 3 above)
Comment (noordawod) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:272]:
```suggestion
		if (old != null) {
			return old;
		}

		return dao;
```

File: src/main/java/com/j256/ormlite/dao/DaoManager.java (hunk 2 above)
Comment (noordawod) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:100]: Keep the suppression please, this is meant for IDEs/linters.

**Review by noordawod — CHANGES_REQUESTED** [review]
The pr looks promising, left a few comments for your consideration. Thank you for your contribution! 🎉

File: src/main/java/com/j256/ormlite/dao/DaoManager.java (hunk 4 above)
Comment (j256) [reply src/main/java/com/j256/ormlite/dao/DaoManager.java:248]: Actually I like the returns at the same level in this case.  What I would ration is if old == null return dao else return old.

File: src/main/java/com/j256/ormlite/dao/DaoManager.java (hunk 3 above)
Comment (j256) [reply src/main/java/com/j256/ormlite/dao/DaoManager.java:272]: Same comment as above.

**Comment by j256** [comment]
My initial response was that we needed to keep the maps in sync so I didn't see a problem with the synchronized.  But we are only dealing with one map at a time so this code is fine.  Thanks for it.

File: src/main/java/com/j256/ormlite/dao/BaseDaoImpl.java (hunk 1 above)
Comment (MarcMil) [reply src/main/java/com/j256/ormlite/dao/BaseDaoImpl.java:60]: I'll undo it. But I was wondering why this was abstract in the first place.
Since it is abstract, one cannot simply instantiate it, but instead there are lots of anonymous classes:
```
new BaseDaoImpl<ForeignFoo, Integer>(connectionSource, ForeignFoo.class) {
				}
```

Why is that?

File: src/main/java/com/j256/ormlite/dao/DaoManager.java (hunk 4 above)
Comment (noordawod) [reply src/main/java/com/j256/ormlite/dao/DaoManager.java:248]: Ahh, I follow the "bail out early" design pattern, explained here: https://medium.com/swlh/return-early-pattern-3d18a41bba8

File: src/main/java/com/j256/ormlite/dao/BaseDaoImpl.java (hunk 1 above)
Comment (noordawod) [reply src/main/java/com/j256/ormlite/dao/BaseDaoImpl.java:60]: Just because this pr doesn't require this change, and because it's untested and it's unclear what kind of repercussions it'll have in the wild. It's better to create a separate pr which introduces this change as part of a bigger refactor to remove abstract classes, for example.

Comment (MarcMil) [reply src/main/java/com/j256/ormlite/dao/BaseDaoImpl.java]: That's true. I've reverted this change.

File: src/main/java/com/j256/ormlite/dao/DaoManager.java (hunk 4 above)
Comment (MarcMil) [reply src/main/java/com/j256/ormlite/dao/DaoManager.java]: Oh, ok. So since it's j256's project, I'll redo with returns on the same level, alright?

**Review by noordawod — APPROVED** [review]
I like this pr and the comments were addressed. Although I haven't tested the changes, I do believe that they're sane and won't interfere with the normal operation of the library.

File: src/main/java/com/j256/ormlite/dao/DaoManager.java [hunk 5]
```diff
@@ -142,11 +142,11 @@ public class DaoManager {
 		Dao<?, ?> dao = lookupDao(key);
 		if (dao == null) {
 			return null;
-		} else {
-			@SuppressWarnings("unchecked")
-			D castDao = (D) dao;
-			return castDao;
 		}
+
```
Comment (j256) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:146]: Sorry can we revert this change?  I see no value in it.

File: src/main/java/com/j256/ormlite/dao/DaoManager.java [hunk 6]
```diff
@@ -163,19 +163,20 @@ public class DaoManager {
 	 * <b>NOTE:</b> You should maybe use the {@link DatabaseTable#daoClass()} and have the DaoManager construct the DAO
 	 * if possible.
 	 * </p>
+	 * @return
```
Comment (j256) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:166]: Remove worthless javadoc.

File: src/main/java/com/j256/ormlite/dao/DaoManager.java [hunk 7]
```diff
@@ -213,57 +214,43 @@ public static synchronized void registerDaoWithTableConfig(ConnectionSource conn
...
 	 */
-	public static synchronized void addCachedDatabaseConfigs(Collection<DatabaseTableConfig<?>> configs) {
-		Map<Class<?>, DatabaseTableConfig<?>> newMap;
-		if (configMap == null) {
-			newMap = new HashMap<Class<?>, DatabaseTableConfig<?>>();
-		} else {
-			newMap = new HashMap<Class<?>, DatabaseTableConfig<?>>(configMap);
-		}
+	public static void addCachedDatabaseConfigs(Collection<DatabaseTableConfig<?>> configs) {
+		Map<Class<?>, DatabaseTableConfig<?>> newMap = new HashMap<Class<?>, DatabaseTableConfig<?>>(configMap);
```
Comment (j256) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:235]: We don't need this newMap.  I would just to a configMap.put in the loop.

File: src/main/java/com/j256/ormlite/dao/DaoManager.java [hunk 8]
```diff
@@ -277,35 +264,21 @@ private static void removeDaosFromConnectionClassMap(ConnectionSource connection
 		}
 	}
 
-	private static void addDaoToTableMap(TableConfigConnectionSource key, Dao<?, ?> dao) {
-		if (tableConfigMap == null) {
-			tableConfigMap = new HashMap<TableConfigConnectionSource, Dao<?, ?>>();
+	private static Dao<?, ?> addDaoToTableMap(TableConfigConnectionSource key, Dao<?, ?> dao) {
```
Comment (j256) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:267]: Change to maybeAdd... or something because it might not add it.

File: src/main/java/com/j256/ormlite/dao/DaoManager.java [hunk 9]
```diff
@@ -213,57 +214,43 @@ public static synchronized void registerDaoWithTableConfig(ConnectionSource conn
...
 			logger.info("Loaded configuration for {}", config.getDataClass());
 		}
-		configMap = newMap;
+		configMap.putAll(newMap);
 	}
 
-	private static void addDaoToClassMap(ClassConnectionSource key, Dao<?, ?> dao) {
-		if (classMap == null) {
-			classMap = new HashMap<ClassConnectionSource, Dao<?, ?>>();
+	private static Dao<?, ?> addDaoToClassMap(ClassConnectionSource key, Dao<?, ?> dao) {
```
Comment (j256) [inline src/main/java/com/j256/ormlite/dao/DaoManager.java:243]: Change to maybeAdd... because it might not add it.

**Review by j256 — COMMENTED** [review]
Thanks.

**Comment by j256** [comment]
And I forget that ormlite hasn't officially moved to Java 8 which is when putIfAbsent() was introduced.  Ok, it's past time.  We will do that and put out a version 7.0.
